python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -944,41 +972,119 @@ def _get_entities_from_payload(self, data):
+
+            entity["project"] = project_entity
             return (project_entity, [entity])
         elif "entity_ids" in data:
+            # Multiple entities were passed down from the web app. This is an older
+            # paradigm, but is still supported. We don't know, however, whether we
+            # got a list of entity ids or a list of entity dictionaries. Old, not-as-
+            # old, and new code do different things, unfortunately. We'll handle all
+            # possible cases pretty easily, though, which will cover current and past
+            # versions of the Shotgun web application.
             entities = []
 
             for entity in data["entity_ids"]:
                 # Did we get an entity list, or a list of entity ids?
                 if isinstance(entity, dict):
-                    if "project" not in entity:
-                        entity["project"] = project_entity
+                    # If we were passed a usable project entity from the web app, we
+                    # can trust that and add it to our entity. If we didn't, then we'll
+                    # have to query it.
+                    if entity.get("project") is None:
+                        if project_entity is None:
+                            entity["project"] = self._get_entity_parent_project(entity)
+                        else:
+                            entity["project"] = project_entity
 
                     entities.append(entity)
                 else:
-                    entities.append(
-                        dict(
-                            type=data["entity_type"],
-                            id=entity,
-                            project=project_entity,
-                        )
+                    entity = dict(
+                        type=data["entity_type"],
+                        id=entity,
                     )
+                    # If we were passed a usable project entity from the web app, we
+                    # can trust that and add it to our entity. If we didn't, then we'll
+                    # have to query it.
+                    if project_entity is None:
+                        entity["project"] = project_entity
+                    else:
+                        entity["project"] = self._get_entity_parent_project(entity)
+
+                    entities.append(entity)
+
+            # If we were not given a usable project entity, we can pull it from an
+            # entity we've just extracted. This doesn't cover the case of receiving
+            # multiple entities from the web app from different projects, but this isn't
+            # the only place where we're going to suffer there.
+            if project_entity is None:
+                project_entity = entities[0]["project"]
+                if project_entity is None:
+                    raise RuntimeError("Unable to determine a project entity from data: %s" % data)
```

Comment:
Is this likely to happen? If so, could you give us an example why it might happen?

Reply:
Should be unlikely verging on impossible. My only thought here was catching some case of somehow getting a site-level entity that doesn't have a project association, and dying here in some way that gives a decent error message. It's just being (overly) cautious, though.